src/dft/backends/portfft/commit.cpp

```diff
@@ -0,0 +1,312 @@
+* Copyright Codeplay Software Ltd
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+#include <array>
+#include <optional>
+
+#include <portfft/portfft.hpp>
+
+#include "oneapi/mkl/exceptions.hpp"
+
+#include "oneapi/mkl/dft/detail/commit_impl.hpp"
+#include "oneapi/mkl/dft/detail/descriptor_impl.hpp"
+#include "oneapi/mkl/dft/detail/portfft/onemkl_dft_portfft.hpp"
+#include "oneapi/mkl/dft/types.hpp"
+
+#include "portfft_helper.hpp"
+
+// alias to avoid ambiguity
+namespace pfft = portfft;
+
+namespace oneapi::mkl::dft::portfft {
+namespace detail {
+
+/// Commit impl class specialization for cuFFT.
+template <dft::precision prec, dft::domain dom>
+class portfft_commit final : public dft::detail::commit_impl<prec, dom> {
+private:
+    using scalar_type = typename dft::detail::commit_impl<prec, dom>::scalar_type;
+    using fwd_type = typename dft::detail::commit_impl<prec, dom>::fwd_type;
+    using bwd_type = typename dft::detail::commit_impl<prec, dom>::bwd_type;
+    using descriptor_type = typename dft::detail::descriptor<prec, dom>;
+
+    static constexpr pfft::domain domain =
+        dom == dft::domain::REAL ? pfft::domain::REAL : pfft::domain::COMPLEX;
+    // since only complex-to-complex transforms are supported, we expect both directions to be valid or neither.
+    std::array<storage_type<descriptor_type>, 2> committed_descriptors = { std::nullopt,
+                                                                           std::nullopt };
+
+public:
+    portfft_commit(sycl::queue& queue, const dft::detail::dft_values<prec, dom>&)
+            : oneapi::mkl::dft::detail::commit_impl<prec, dom>(queue, backend::portfft) {
+        if constexpr (prec == dft::detail::precision::DOUBLE) {
+            if (!queue.get_device().has(sycl::aspect::fp64)) {
+                throw mkl::exception("DFT", "commit", "Device does not support double precision.");
+            }
+        }
+    }
+
+    void commit(const dft::detail::dft_values<prec, dom>& config_values) override {
+        // not available in portFFT:
+        if (config_values.workspace != config_value::ALLOW) {
+            throw mkl::unimplemented("dft/backends/portfft", __FUNCTION__,
```

Comment:
Maybe it is already done and I missed it but it seems that portFFT support only C2C FFT. So can we make sure to throw an unimplemented  exception if user try to compute R2C or C2R FFT with portFFT backend?

Reply:
During commit, portFFT will throw a `portfft::unsupported_configuration` exception, which will be caught and `oneapi::mkl::unimplemented` will be thrown with the details.